### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by emilcheva**
Mocked graphql queries with MSW and added tests accordingly

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -51,14 +51,16 @@
...
+    "@testing-library/user-event": "^13.5.0",
     "autoprefixer": "^10.4.1",
     "eslint-config-prettier": "^8.3.0",
     "eslint-plugin-prettier": "^4.0.0",
     "lodash.debounce": "^4.0.8",
+    "msw": "^0.36.7",
     "postcss": "^8.4.5",
-    "prettier": "^2.5.1"
+    "prettier": "^2.5.1",
+    "react-test-renderer": "^17.0.2"
```
Comment (steliyan) [inline package.json:63]: NIT: I would lock the dependency versions. If you decide to do it, do it in a separate PR.

File: src/ApolloClient.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
+
+const cache = new InMemoryCache();
+
+const link = new HttpLink({
+  uri: 'https://cat-wiki-apollo-server.herokuapp.com',
+  cache: new InMemoryCache()
+});
```
Comment (steliyan) [inline src/ApolloClient.js]: Why 2 separate cache objects? It's bound to end in some weird UI state error.

File: src/components/__tests__/autocomplete.test.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import { ApolloProvider } from '@apollo/client';
+import { fireEvent, render, screen, waitFor } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import { client } from '../../ApolloClient';
+import { setupServer } from '../../mocks/setupServer';
+import Autocomplete from '../autocomplete';
+
+const autoCompleteSetup = () => {
+  render(
+    <ApolloProvider client={client}>
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: The way the client is passed to the `ApolloProvider` the `client` is shared between the tests in this file (test suite). Is this expected?

File: src/mocks/setupServer.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { server } from './server';
+
+export const setupServer = function () {
```
Comment (steliyan) [inline src/mocks/setupServer.js]: Why not an arrow function?

File: src/mocks/handlers.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
+import { graphql } from 'msw';
+
+export const handlers = [
+  // Handles a "GetBreeds" query
+  graphql.query('getBreeds', (req, res, ctx) => {
+    // const { limit } = req.variables;
```
Comment (steliyan) [inline src/mocks/handlers.js]: Leftover?

File: src/components/autocomplete.js [hunk 6]
```diff
@@ -8,7 +8,7 @@ import styled from '@emotion/styled';
 import { colors } from '../styles';
 
 export const GET_BREED = gql`
-  query getBreed($breedName: String!) {
+  query getAutoCompleteBreed($breedName: String!) {
```
Comment (steliyan) [inline src/components/autocomplete.js:11]: Why not `getBreedByName`?

File: src/components/autocomplete.js [hunk 7]
```diff
@@ -26,18 +26,18 @@ const Autocomplete = () => {
...
-  function changeHandler(query) {
-    if (!query) return setSuggestions([]);
-    setInput(query);
-
-    const debouncedChangeFilter = debounce(() => {
+  const debouncedApiCall = useCallback(
+    debounce((text) => {
       getCatBreeds({
-        variables: { breedName: query }
+        variables: { breedName: text }
```
Comment (steliyan) [inline src/components/autocomplete.js]: NIT: You can rename `text` to `breedName` and simplify the obj construction.

File: src/components/autocomplete.js [hunk 8]
```diff
@@ -26,18 +26,18 @@ const Autocomplete = () => {
...
-    }, 500);
+    }, 500),
+    []
+  );
 
-    debouncedChangeFilter();
-  }
+  useEffect(() => {
+    debouncedApiCall(input);
+  }, [input, debouncedApiCall]);
```
Comment (steliyan) [inline src/components/autocomplete.js]: It's a good idea to clean-up in case this component gets unmounted.

You can use `.cancel`: https://lodash.com/docs/4.17.15#debounce

File: src/components/autocomplete.js [hunk 9]
```diff
@@ -26,18 +26,18 @@ const Autocomplete = () => {
...
 
   useEffect(() => {
     if (data && !loading) {
```
Comment (steliyan) [inline src/components/autocomplete.js:47]: I think this may be true if the current query returns an error, then `data` would hold the previous result. Not sure if this is what we want.

File: src/components/autocomplete.js [hunk 10]
```diff
@@ -46,16 +46,10 @@ const Autocomplete = () => {
...
 
-  useEffect(() => {
-    if (!suggestions.length) {
-      setInput('');
-    }
-  }, [suggestions]);
-
   const keyDownHandler = (e) => {
     // enter key
     if (e.keyCode === 13) {
```
Comment (steliyan) [inline src/components/autocomplete.js:56]: `const ENTER_KEY_CODE = 13` and the comment can safely be deleted. :)

File: src/components/breed-detail.js [hunk 11]
```diff
@@ -31,7 +31,16 @@ const BreedDetail = ({ breed }) => {
...
+            {breedImage.length ? (
+              <ImageWrapper width="270" height="270" src={breedImage[0]?.url} alt={name} />
+            ) : (
+              <ImageWrapper
+                width="270"
+                height="270"
+                src="https://images.unsplash.com/photo-1577349516274-37ff88a53627?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1035&q=80"
+                alt={name}
+              />
+            )}
```
Comment (steliyan) [inline src/components/breed-detail.js]: Both `ImageWrapper`s are basically the same, why not just specify the appropriate prop?

File: src/components/breed-detail.js [hunk 12]
```diff
@@ -77,14 +86,26 @@ const BreedDetail = ({ breed }) => {
...
+                    width="100%"
+                    height="250"
+                    alt={name}
+                  />
+                </Col>
+              ))}
+            </Row>
+          </>
+        ) : (
+          ''
```
Comment (steliyan) [inline src/components/breed-detail.js]: Why not `nuill`?

File: src/components/breed-detail.js [hunk 13]
```diff
@@ -77,14 +86,26 @@ const BreedDetail = ({ breed }) => {
...
-              <ImageWrapper key={image.id} src={image.url} width="100%" height="250" alt={name} />
-            </Col>
-          ))}
-        </Row>
+        {breedImage.length ? (
+          <>
+            <h2>Other photos</h2>
+            <Row>
+              {breedImage.map((image, index) => (
+                <Col sm="6" lg="3" className="mb-4" key={index}>
```
Comment (steliyan) [inline src/components/breed-detail.js]: Don't use `index` for a `key`: https://reactjs.org/docs/lists-and-keys.html

File: src/mocks/setupServer.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+  beforeAll(() => {
+    server.listen({
+      onUnhandledRequest: 'warn'
```
Comment (steliyan) [inline src/mocks/setupServer.js]: Why is that needed? I would advise treating every warning as an error in the tests as this may uncover potential problems with the production code (or the mocks).

File: src/components/__tests__/autocomplete.test.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+      <Autocomplete />
+    </ApolloProvider>
+  );
+};
+
+describe('Testing Autocomplete w getBreedsByName query', () => {
+  setupServer();
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: This gets executed once per `describe`. Most likely, you want to do it before each test. If not, `beforeAll` can be used scoped in the `describe`. Either way, it's more explicit and easier for the reader.

File: src/components/__tests__/autocomplete.test.js [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+  it('should render Autocomplete results when text is entered', async () => {
+    autoCompleteSetup();
+    userEvent.type(screen.getByPlaceholderText(/enter your breed/i), 'po');
+
+    await waitFor(() => {
+      expect(screen.queryAllByTestId('suggestions-list')[0]).toBeInTheDocument();
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js:45]: Seems redundant given the assertion below.

File: src/components/__tests__/autocomplete.test.js [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    });
+    screen.getByPlaceholderText(/enter your breed/i).focus();
+    fireEvent.keyDown(document.activeElement || document.body, {
+      key: 'Enter',
+      keyCode: 13,
+      which: 13
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: I think `shich` is redundant here.

File: src/components/__tests__/autocomplete.test.js [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+  it('should select first breed name on enter', async () => {
+    autoCompleteSetup();
+    userEvent.type(screen.getByPlaceholderText(/enter your breed/i), 'po');
+
+    await waitFor(() => {
+      expect(screen.queryAllByTestId('suggestions-list')[0]).toBeInTheDocument();
+    });
+    screen.getByPlaceholderText(/enter your breed/i).focus();
+    fireEvent.keyDown(document.activeElement || document.body, {
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: Why the `||` here?

File: src/components/__tests__/components.test.js [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    render(
+      <ApolloProvider client={client}>
+        <Breed breedName="Munchkin" />
+      </ApolloProvider>
+    );
+
+    const breedName = await screen.findByText('Munchkin');
+    expect(breedName).toBeTruthy();
+    const temperament = await screen.findByText(/agile, easy going, intelligent, playful/i);
+    expect(temperament).toBeTruthy();
```
Comment (steliyan) [inline src/components/__tests__/components.test.js]: Can we use `toBeInTheDocument` instead of `toBeTruthy`?

File: src/components/__tests__/components.test.js [hunk 20]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import { ApolloProvider } from '@apollo/client';
+import { render, screen } from '@testing-library/react';
+import { client } from '../../ApolloClient';
+import TopBreeds from '../../pages/top-breeds';
+import { setupServer } from '../../mocks/setupServer';
+import Breed from '../../pages/breed';
+
+describe('Testing TopBreeds w getBreeds query', () => {
```
Comment (steliyan) [inline src/components/__tests__/components.test.js]: `Testing` in the title, why? :)

File: src/components/__tests__/query-result.test.js [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+  });
+
+  it('renders No Data message', async () => {
+    render(<QueryResult loading={false} />);
+    expect(screen.getByText(/no data to show/i)).toBeInTheDocument();
+  });
+
+  it('renders Error', async () => {
+    render(<QueryResult error={new ApolloError('something erroneous happened')} />);
+    expect(screen.getByText((content) => content.startsWith('ERROR'))).toBeInTheDocument();
```
Comment (steliyan) [inline src/components/__tests__/query-result.test.js]: No need for `async`.

File: src/components/__tests__/query-result.test.js [hunk 22]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { render, cleanup, screen } from '@testing-library/react';
+import QueryResult from '../query-result';
```
Comment (steliyan) [inline src/components/__tests__/query-result.test.js:3]: That's a weird component, given how you use it. Why the null-check in the children?

File: src/components/__tests__/autocomplete.test.js (hunk 3 above)
Comment (emilcheva) [reply src/components/__tests__/autocomplete.test.js]: yep, intended :)

File: package.json (hunk 1 above)
Comment (emilcheva) [reply package.json:63]: Ok, will resolve with another PR

File: src/ApolloClient.js (hunk 2 above)
Comment (emilcheva) [reply src/ApolloClient.js]: Fixed

File: src/ApolloClient.js [hunk 23]
```diff
@@ -4,7 +4,7 @@ const cache = new InMemoryCache();
 
 const link = new HttpLink({
   uri: 'https://cat-wiki-apollo-server.herokuapp.com',
-  cache: new InMemoryCache()
+  cache: cache
```
Comment (steliyan) [inline src/ApolloClient.js]: `cache: cache` -> `cache`.

File: src/components/autocomplete.js [hunk 24]
```diff
@@ -63,7 +68,7 @@ const Autocomplete = () => {
     }
     // down arrow
     else if (e.keyCode === 40) {
```
Comment (steliyan) [inline src/components/autocomplete.js:70]: Same comment as for the `ENTER` key.

File: src/mocks/handlers.js [hunk 25]
```diff
@@ -73,17 +73,26 @@ export const handlers = [
...
+        })
+      );
+    } else if (breedName === 'brv') {
+      return res(
+        ctx.data({
+          getBreedsByName: []
+        })
+      );
+    }
   })
```
Comment (steliyan) [inline src/mocks/handlers.js]: Should there be `else if` or just `else`?

File: src/components/__tests__/autocomplete.test.js (hunk 3 above)
Comment (steliyan) [reply src/components/__tests__/autocomplete.test.js]: Why?

File: src/components/breed-detail.js [hunk 26]
```diff
@@ -31,16 +31,18 @@ const BreedDetail = ({ breed }) => {
...
-              />
-            )}
+            (
+            <ImageWrapper
+              width="270"
+              height="270"
+              src={
+                breedImage.length
+                  ? breedImage[0]?.url
+                  : 'https://images.unsplash.com/photo-1577349516274-37ff88a53627?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1035&q=80'
```
Comment (steliyan) [inline src/components/breed-detail.js:40]: NIT: You can extract this in a separate func.

File: src/components/breed-detail.js [hunk 27]
```diff
@@ -90,8 +92,8 @@ const BreedDetail = ({ breed }) => {
           <>
             <h2>Other photos</h2>
             <Row>
-              {breedImage.map((image, index) => (
-                <Col sm="6" lg="3" className="mb-4" key={index}>
+              {breedImage.map((image) => (
+                <Col sm="6" lg="3" className="mb-4" key={breedImage.id}>
```
Comment (steliyan) [inline src/components/breed-detail.js]: Should this be `breedImage.id` or `image.id`?

File: src/components/breed-detail.js [hunk 28]
```diff
@@ -90,8 +92,8 @@ const BreedDetail = ({ breed }) => {
           <>
```
Comment (steliyan) [inline src/components/breed-detail.js:90]: Does the `.length` condition work if `0` items exist?

File: src/components/autocomplete.js [hunk 29]
```diff
@@ -26,36 +26,35 @@ const Autocomplete = () => {
...
+    []
+  );
 
-    debouncedChangeFilter();
-  }
+  useEffect(() => {
+    debouncedApiCall(input);
+
+    return () => {
+      debouncedApiCall.cancel();
```
Comment (steliyan) [inline src/components/autocomplete.js:42]: Does this work? Shouldn't the return result of the callback get cancelled?

File: src/components/autocomplete.js [hunk 30]
```diff
@@ -82,17 +81,22 @@ const Autocomplete = () => {
...
           onKeyDown={keyDownHandler}
           value={input}
-          aria-label="Recipient's username"
+          aria-label="Search by Breed Name"
           aria-describedby="search"
         />
         <InputSearch id="search">
-          <Link to={`/breed/${input}`}>
-            <Search style={{ fill: colors.dark }} />
+          <Link to={input ? `/breed/${input}` : ''}>
```
Comment (steliyan) [inline src/components/autocomplete.js:93]: Do we need a `Link` component in the `input` is falsy?

File: src/components/autocomplete.js [hunk 31]
```diff
@@ -82,17 +81,22 @@ const Autocomplete = () => {
...
+            <Search
+              style={{
+                fill: input ? colors.dark : colors.darkGrey,
+                cursor: input ? 'pointer' : 'not-allowed'
+              }}
```
Comment (steliyan) [inline src/components/autocomplete.js:98]: The style object is new on each rerender. Generally, we don't use inline styles.

File: src/mocks/handlers.js [hunk 32]
```diff
@@ -0,0 +1,98 @@
+import { graphql } from 'msw';
+
+export const handlers = [
+  // Handles a "GetBreeds" query
```
Comment (steliyan) [inline src/mocks/handlers.js]: Pretty redundant. Most likely will get outdated if something gets fixed down the file.

File: src/mocks/handlers.js [hunk 33]
```diff
@@ -0,0 +1,98 @@
...
+        })
+      );
+    } else if (breedName === 'brv') {
+      return res(
+        ctx.data({
+          getBreedsByName: []
+        })
+      );
+    }
+  })
```
Comment (steliyan) [inline src/mocks/handlers.js]: Same question as before, should this be just `else`?

File: src/components/__tests__/autocomplete.test.js [hunk 34]
```diff
@@ -6,6 +6,7 @@ import { client } from '../../ApolloClient';
 import { setupServer } from '../../mocks/setupServer';
 import { server } from '../../mocks/server';
 import Autocomplete from '../autocomplete';
+import { jssPreset } from '@material-ui/core';
 
 const ENTER_KEY_CODE = 13;
 const REQUESTS_MAP = new Map();
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: That's not really a "const", so better to be `requestsMap`.

File: src/components/__tests__/autocomplete.test.js [hunk 35]
```diff
@@ -111,10 +112,13 @@ describe('Autocomplete w getBreedsByName query', () => {
...
 
-    const getLastKeyInMap = (map) => [...map][map.size - 1][0];
+    let getAllRequestVariables = [];
 
-    expect(REQUESTS_MAP.get(getLastKeyInMap(REQUESTS_MAP))).toStrictEqual({
-      breedName: 'pers'
-    });
+    for (let [, value] of REQUESTS_MAP.entries()) {
+      getAllRequestVariables.push(value['breedName']);
+    }
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: This can be rewritten as:
```const getAllRequestVariables = REQUESTS_MAP.values().map(({ breedName }) => breedName)```.

File: src/components/__tests__/autocomplete.test.js [hunk 36]
```diff
@@ -111,10 +112,13 @@ describe('Autocomplete w getBreedsByName query', () => {
...
+
+    expect(getAllRequestVariables.some((variable) => variable === 'p')).toBe(false);
+    expect(getAllRequestVariables.some((variable) => variable === 'pers')).toBe(true);
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: `toBe(false)` -> `toBeFalsy`
`toBe(true)` -> `toBeTruthy`

Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: If the API call is made twice, the test would pass again.

File: src/components/__tests__/autocomplete.test.js [hunk 37]
```diff
@@ -111,10 +112,13 @@ describe('Autocomplete w getBreedsByName query', () => {
...
+    let getAllRequestVariables = [];
 
-    expect(REQUESTS_MAP.get(getLastKeyInMap(REQUESTS_MAP))).toStrictEqual({
-      breedName: 'pers'
-    });
+    for (let [, value] of REQUESTS_MAP.entries()) {
+      getAllRequestVariables.push(value['breedName']);
+    }
+
+    expect(getAllRequestVariables.some((variable) => variable === 'p')).toBe(false);
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: Why don't we check for `pe` and `per` then?

File: src/components/__tests__/autocomplete.test.js [hunk 38]
```diff
@@ -112,13 +111,11 @@ describe('Autocomplete w getBreedsByName query', () => {
...
-    for (let [, value] of REQUESTS_MAP.entries()) {
-      getAllRequestVariables.push(value['breedName']);
-    }
-
-    expect(getAllRequestVariables.some((variable) => variable === 'p')).toBe(false);
-    expect(getAllRequestVariables.some((variable) => variable === 'pers')).toBe(true);
+    expect(getAllRequestVariables.some((variable) => variable === 'p')).toBeFalsy;
+    expect(getAllRequestVariables.some((variable) => variable === 'pe')).toBeFalsy;
+    expect(getAllRequestVariables.some((variable) => variable === 'per')).toBeFalsy;
+    expect(getAllRequestVariables.some((variable) => variable === 'pers')).toBeTruthy;
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: `toBeFalsy`/`toBeTruthy` are functions, but they are not invoked.

File: src/components/__tests__/autocomplete.test.js [hunk 39]
```diff
@@ -112,13 +111,11 @@ describe('Autocomplete w getBreedsByName query', () => {
       expect(screen.queryAllByTestId('suggestions-list')[0]).toBeInTheDocument();
     });
 
-    let getAllRequestVariables = [];
+    const getAllRequestVariables = [requestsMap.values()].map(({ breedName }) => breedName);
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: Shouldn't the `values()` get spreaded?

File: src/components/__tests__/autocomplete.test.js (hunk 38 above)
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: Can we just do `expect(getAllRequestVariables).toEqual(['pers'])`?

File: src/ApolloClient.js [hunk 40]
```diff
@@ -1,15 +1,13 @@
...
 
-// Isolate Apollo client so it could be reused
-// in both application runtime and tests.
-export const client = new ApolloClient({
-  cache,
-  link
+export const CREATE_CLIENT = new ApolloClient({
+  link,
+  cache
 });
```
Comment (steliyan) [inline src/ApolloClient.js]: This should be a function, not a client instance.

File: src/ApolloClient.js [hunk 41]
```diff
@@ -1,15 +1,13 @@
-import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
+import { HttpLink, InMemoryCache, ApolloClient } from '@apollo/client';
 
 const cache = new InMemoryCache();
 
 const link = new HttpLink({
   uri: 'https://cat-wiki-apollo-server.herokuapp.com',
-  cache: cache
+  cache
```
Comment (steliyan) [inline src/ApolloClient.js]: NIT: You can just inline the cache creation here.

File: src/ApolloClient.js [hunk 42]
```diff
@@ -1,15 +1,13 @@
...
 });
 
-// Isolate Apollo client so it could be reused
-// in both application runtime and tests.
-export const client = new ApolloClient({
-  cache,
-  link
+export const CREATE_CLIENT = new ApolloClient({
+  link,
+  cache
```
Comment (steliyan) [inline src/ApolloClient.js]: Is this needed here?

File: src/components/__tests__/autocomplete.test.js [hunk 43]
```diff
@@ -2,28 +2,40 @@ import React from 'react';
...
+});
+
+afterEach(() => {
+  requestsMap.clear();
+  CREATE_CLIENT.resetStore();
+  server.resetHandlers();
+});
+
+afterAll(() => {
+  server.events.removeAllListeners();
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: Better to remove only the registered listener, not all.

File: src/components/__tests__/autocomplete.test.js [hunk 44]
```diff
@@ -2,28 +2,40 @@ import React from 'react';
...
+beforeAll(() => {
+  server.listen();
+  server.events.on('request:start', (req) => {
+    requestsMap.set(req.id, req.body.variables);
+  });
+});
+
+afterEach(() => {
+  requestsMap.clear();
+  CREATE_CLIENT.resetStore();
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: This would be redundant once the client is instanced per test, not per test suite.

**Review by steliyan — APPROVED** [review]

File: src/components/__tests__/autocomplete.test.js [hunk 45]
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
 
     expect(getAllRequestVariables).toEqual(['pers']);
   });
+
+  it('should not send request for fast typing ( keystroke faster than 500ms )', async () => {
+    jest.useFakeTimers('modern');
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js:133]: This should be default for jest 27.

File: src/components/__tests__/autocomplete.test.js [hunk 46]
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
...
+
+    const searchInput = screen.getByPlaceholderText(/enter your breed/i);
+    let typed = [];
+
+    await userEvent.type(searchInput, 'sphynx');
+
+    act(() => {
+      jest.advanceTimersByTime(499);
+    });
+    expect([...requestsMap.values()].length).toEqual(0);
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: You can use `toHaveLength`:  https://jestjs.io/docs/expect#tohavelengthnumber

File: src/components/__tests__/autocomplete.test.js [hunk 47]
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
 
     expect(getAllRequestVariables).toEqual(['pers']);
   });
+
+  it('should not send request for fast typing ( keystroke faster than 500ms )', async () => {
+    jest.useFakeTimers('modern');
+    autoCompleteSetup();
+
+    const searchInput = screen.getByPlaceholderText(/enter your breed/i);
+    let typed = [];
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: Do we need this?

File: src/components/__tests__/autocomplete.test.js [hunk 48]
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
...
+    act(() => {
+      jest.advanceTimersByTime(499);
+    });
+    expect([...requestsMap.values()].length).toEqual(0);
+    act(() => {
+      jest.advanceTimersByTime(501);
+      typed.push(searchInput.value);
+    });
+    const getAllRequestVariables = [...requestsMap.values()].map(({ breedName }) => breedName);
+    expect(getAllRequestVariables).toEqual([...typed]);
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: Why the spread here?

File: src/components/__tests__/autocomplete.test.js [hunk 49]
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
...
+    let typed = [];
+
+    await userEvent.type(searchInput, 'sphynx');
+
+    act(() => {
+      jest.advanceTimersByTime(499);
+    });
+    expect([...requestsMap.values()].length).toEqual(0);
+    act(() => {
+      jest.advanceTimersByTime(501);
```
Comment (steliyan) [inline src/components/__tests__/autocomplete.test.js]: Should this be `501` or just `1`?
